python-package/lightgbm/basic.py

```diff
@@ -240,7 +240,7 @@ def _is_numpy_column_array(data: Any) -> bool:
     return len(shape) == 2 and shape[1] == 1
 
 
-def _cast_numpy_array_to_dtype(array: np.ndarray, dtype: np.dtype) -> np.ndarray:
+def _cast_numpy_array_to_dtype(array: np.ndarray, dtype: "np.typing.DTypeLike") -> np.ndarray:
```

Comment:
these annotations come from https://numpy.org/devdocs/reference/typing.html.

`numpy.typing` is new as of `numpy` 1.20 ([January 2021](https://pypi.org/project/numpy/#history)).

Expressing these annotations as string literals means that they won't cause issues at runtime for users using `lightgbm` + older `numpy` versions. See https://mypy.readthedocs.io/en/stable/runtime_troubles.html#import-cycles.